pkg/distributor/distributor_test.go

```diff
@@ -1426,63 +1426,78 @@ func TestDistributor_Push_ExemplarValidation(t *testing.T) {
-			errID:  globalerror.ExemplarLabelsTooLong,
+		"drops exemplar with too long labelset": {
+			req:     makeWriteRequestExemplar([]string{model.MetricNameLabel, "test"}, 1000, 1, []string{"foo", strings.Repeat("0", 126)}),
+			expDrop: true,
 		},
 		"rejects exemplar with too many series labels": {
-			req:    makeWriteRequestExemplar(manyLabels, 0, nil),
-			errMsg: "received a series whose number of labels exceeds the limit",
-			errID:  globalerror.MaxLabelNamesPerSeries,
+			req:       makeWriteRequestExemplar(manyLabels, 0, 1, nil),
+			expErrMsg: "received a series whose number of labels exceeds the limit",
+			expErrID:  globalerror.MaxLabelNamesPerSeries,
 		},
 		"rejects exemplar with duplicate series labels": {
-			req:    makeWriteRequestExemplar([]string{model.MetricNameLabel, "test", "foo", "bar", "foo", "bar"}, 0, nil),
-			errMsg: "received a series with duplicate label name",
-			errID:  globalerror.SeriesWithDuplicateLabelNames,
+			req:       makeWriteRequestExemplar([]string{model.MetricNameLabel, "test", "foo", "bar", "foo", "bar"}, 0, 1, nil),
+			expErrMsg: "received a series with duplicate label name",
+			expErrID:  globalerror.SeriesWithDuplicateLabelNames,
 		},
 		"rejects exemplar with empty series label name": {
-			req:    makeWriteRequestExemplar([]string{model.MetricNameLabel, "test", "", "bar"}, 0, nil),
-			errMsg: "received a series with an invalid label",
-			errID:  globalerror.SeriesInvalidLabel,
+			req:       makeWriteRequestExemplar([]string{model.MetricNameLabel, "test", "", "bar"}, 0, 1, nil),
+			expErrMsg: "received a series with an invalid label",
+			expErrID:  globalerror.SeriesInvalidLabel,
 		},
 	}
 
 	for testName, tc := range tests {
 		t.Run(testName, func(t *testing.T) {
+			expSamples := tc.req.Timeseries[0].Samples
+			expExemplars := tc.req.Timeseries[0].Exemplars
 			limits := prepareDefaultLimits()
 			limits.MaxGlobalExemplarsPerUser = 10
-			ds, _, _, _ := prepare(t, prepConfig{
+			ds, ingesters, _, _ := prepare(t, prepConfig{
 				limits:           limits,
 				numIngesters:     2,
 				happyIngesters:   2,
 				numDistributors:  1,
 				shuffleShardSize: 0,
 			})
 			_, err := ds[0].Push(ctx, tc.req)
-			if tc.errMsg != "" {
+			if tc.expErrMsg != "" {
+				require.Error(t, err)
 				fromError, _ := grpcutil.ErrorToStatus(err)
-				assert.Contains(t, fromError.Message(), tc.errMsg)
-				assert.Contains(t, fromError.Message(), tc.errID)
-			} else {
-				assert.Nil(t, err)
+				assert.Contains(t, fromError.Message(), tc.expErrMsg)
+				assert.Contains(t, fromError.Message(), tc.expErrID)
+				return
+			}
+
+			require.NoError(t, err)
+			for _, i := range ingesters {
+				ss := i.series()
+				require.Len(t, ss, 1)
+				for _, s := range ss {
+					require.Equal(t, expSamples, s.Samples)
+					if !tc.expDrop {
+						// Not sure why, but during the push, the request's exemplar labels become empty
```

Comment:
Seems bad that we don't know why something is behaving this way if we're trying to test it.

Reply:
I'll see if I can find out, but it's mysterious. I'm guessing the mutation happens somewhere due to some performance tweak.

Reply:
To be clear though, this behaviour is orthogonal to what we're fixing/testing. It's not the result that's affected, but the request being passed in.

Reply:
My bet is that we're sending the input of testcase as the write request, but `Push()` does the request cleanup returning back all the items from the write request to the pools, zero-ing the labels of everything in that process to avoid retaining references to the unmarshaled input bytes.

Even though it's unrelated to what you're fixing, it's a good moment to fix that: instead of passing the request's input, you should copy it (you can also marshal it to bytes and then unmarshal).

Reply:
Thanks for pointing me in the right direction @colega :)

Reply:
Done, PTAL.

Reply:
Thanks, I left a nit suggestion in another thread, but LGTM.